### PR TITLE
Task lwdev 8693 cashout

### PR DESCRIPTION
**Pull request by mmusin**

### CODE REVIEW

File: src/Lykke.Service.Operations/Controllers/OperationsController.cs [hunk 1]
```diff
@@ -409,23 +456,37 @@ public async Task Confirm(Guid id)
             if (operation.Status == OperationStatus.Confirmed)
                 return;
 
-            if (operation.Status != OperationStatus.Created) // todo: accepted?
+            if (operation.Status != OperationStatus.Created && operation.Status != OperationStatus.Accepted) // todo: accepted?
```
Comment (polonzer) [inline src/Lykke.Service.Operations/Controllers/OperationsController.cs]: "todo" is implemented here.

File: src/Lykke.Service.Operations/Modules/ServiceModule.cs [hunk 2]
```diff
@@ -1,51 +1,66 @@
...
                 .SingleInstance();
-            
-            builder.RegisterInstance<ILimitOrdersRepository>(
-                new LimitOrdersRepository(AzureTableStorage<LimitOrderEntity>.Create(_settings.ConnectionString(x => x.OperationsService.Db.HMarketOrdersConnString),
-                    "LimitOrders", _log)));
+
+            var redis = new RedisCache(new RedisCacheOptions
+            {
+                Configuration = _settings.CurrentValue.RedisSettings.Configuration,
+                InstanceName = "FinanceDataCacheInstance"
```
Comment (polonzer) [inline src/Lykke.Service.Operations/Modules/ServiceModule.cs]: Why do we want to use this prefix?

File: src/Lykke.Service.Operations/Services/ProtoSerializer.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
+﻿using System.IO;
+using Lykke.RabbitMqBroker.Subscriber;
+
+namespace Lykke.Service.Operations.Services
+{
+    public class ProtoSerializer<TMessage> : IMessageDeserializer<TMessage>
```
Comment (polonzer) [inline src/Lykke.Service.Operations/Services/ProtoSerializer.cs]: As far as I see, this class is not used.

File: src/Lykke.Service.Operations/Workflow/CommandHandlers/CommandHandler.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        {
+            _log = logFactory.CreateLog(this);
+            _operationsRepository = operationsRepository;
+            _ethereumHotWallet = ethereumHotWallet;
+        }
+
+        [UsedImplicitly]
+        public async Task<CommandHandlingResult> Handle(CreateCashoutCommand command, IEventPublisher eventPublisher)
+        {
+            _log.Info($"CreateCashoutCommand received. Operation [{command.OperationId}]", command);
```
Comment (polonzer) [inline src/Lykke.Service.Operations/Workflow/CommandHandlers/CommandHandler.cs]: Isn't it excessive logging?

File: src/Lykke.Service.Operations/Workflow/CommandHandlers/SolarCoinCommandHandler.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+        public SolarCoinCommandHandler(ILogFactory logFactory, ISrvSolarCoinCommandProducer solarCoinCommandProducer)
+        {
+            _log = logFactory.CreateLog(this);
+            _solarCoinCommandProducer = solarCoinCommandProducer;
+        }
+
+        public async Task<CommandHandlingResult> Handle(SolarCashOutCommand command, IEventPublisher eventPublisher)
+        {
+            _log.WriteInfo(nameof(SolarCoinCommandHandler), command.ToJson(), "SolarCashOutCommand received");
```
Comment (polonzer) [inline src/Lykke.Service.Operations/Workflow/CommandHandlers/SolarCoinCommandHandler.cs]: Isn't it excessive logging?

File: src/Lykke.Service.Operations/Workflow/CommandHandlers/WorkflowCommandHandler.cs [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+            _log = logFactory.CreateLog(this);
+            _operationsRepository = operationsRepository;
+            _workflowService = workflowService;
+            _workflowFactory = workflowFactory;
+        }
+
+        [UsedImplicitly]
+        public async Task<CommandHandlingResult> Handle(ExecuteOperationCommand command, IEventPublisher eventPublisher)
+        {
+            _log.Info($"ExecuteOperationCommand received. Operation [{command.OperationId}]", command);
```
Comment (polonzer) [inline src/Lykke.Service.Operations/Workflow/CommandHandlers/WorkflowCommandHandler.cs]: Isn't it excessive logging?

File: src/Lykke.Service.Operations/Workflow/CommandHandlers/WorkflowCommandHandler.cs [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+
+            await HandleWorkflow(operation, wfResult, eventPublisher);
+
+            return CommandHandlingResult.Ok();
+        }
+
+        [UsedImplicitly]
+        public async Task<CommandHandlingResult> Handle(CompleteActivityCommand command, IEventPublisher eventPublisher)
+        {
+            _log.Info($"CompleteActivityCommand received. Operation [{command.OperationId}]", command);
```
Comment (polonzer) [inline src/Lykke.Service.Operations/Workflow/CommandHandlers/WorkflowCommandHandler.cs]: Isn't it excessive logging?

File: src/Lykke.Service.Operations/Workflow/CommandHandlers/WorkflowCommandHandler.cs [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+            if (wfResult != null)
+                await HandleWorkflow(operation, wfResult, eventPublisher);
+
+            return CommandHandlingResult.Ok();
+        }
+
+        [UsedImplicitly]
+        public async Task<CommandHandlingResult> Handle(FailActivityCommand command, IEventPublisher eventPublisher)
+        {
+            _log.Info($"FailActivityCommand received. Operation [{command.OperationId}]", command);
```
Comment (polonzer) [inline src/Lykke.Service.Operations/Workflow/CommandHandlers/WorkflowCommandHandler.cs]: Isn't it excessive logging?

File: src/Lykke.Service.Operations/Workflow/Commands/ExecuteOperationCommand.cs [hunk 9]
```diff
@@ -0,0 +1,9 @@
+﻿using System;
+
+namespace Lykke.Service.Operations.Workflow.Commands
+{
+    public class ExecuteOperationCommand
```
Comment (polonzer) [inline src/Lykke.Service.Operations/Workflow/Commands/ExecuteOperationCommand.cs]: [MessagePackObject] ?
In fact, I think we can just use default MessagePack serialization settings (and set them, if needed).

File: src/Lykke.Service.Operations/Workflow/Sagas/BlockchainCashoutSaga.cs [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+
+                    _log.Info($"StartCashoutCommand for Bitcoin has sent. Operation [{command.Id}]", command);
+                }
+            }
+        }        
+
+        [UsedImplicitly]
+        public async Task Handle(Bitcoin.Contracts.Events.CashoutCompletedEvent evt, ICommandSender commandSender)
+        {
+            _log.Info($"CashoutCompletedEvent for Bitcoin received. Operation [{evt.OperationId}]", evt);
```
Comment (polonzer) [inline src/Lykke.Service.Operations/Workflow/Sagas/BlockchainCashoutSaga.cs]: Isn't it excessive logging?

File: src/Lykke.Service.Operations/Workflow/Sagas/BlockchainCashoutSaga.cs [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+                }.ToJson()
+            };
+
+            commandSender.SendCommand(command, "operations");
+        }
+
+        [UsedImplicitly]
+        public async Task Handle(Job.EthereumCore.Contracts.Cqrs.Events.CashoutCompletedEvent evt, ICommandSender commandSender)
+        {
+            _log.Info($"CashoutCompletedEvent for Ethereum received. Operation [{evt.OperationId}]", evt);
```
Comment (polonzer) [inline src/Lykke.Service.Operations/Workflow/Sagas/BlockchainCashoutSaga.cs]: Isn't it excessive logging?

File: src/Lykke.Service.Operations/Workflow/Sagas/BlockchainCashoutSaga.cs [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+                Output = "{}"
+            };
+
+            commandSender.SendCommand(command, "operations");
+        }
+
+        [UsedImplicitly]
+        public async Task Handle(SolarCashOutCompletedEvent evt, ICommandSender commandSender)
+        {
+            _log.Info($"CashoutCompletedEvent for Solarcoin received. Operation [{evt.OperationId}]", evt);
```
Comment (polonzer) [inline src/Lykke.Service.Operations/Workflow/Sagas/BlockchainCashoutSaga.cs]: Isn't it excessive logging?

File: src/Lykke.Service.Operations/Workflow/Sagas/BlockchainCashoutSaga.cs [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+                Output = "{}"
+            };
+
+            commandSender.SendCommand(command, "operations");
+        }
+        
+        [UsedImplicitly]
+        public async Task Handle(CashoutCompletedEvent evt, ICommandSender commandSender)
+        {
+            _log.Info($"CashoutCompletedEvent for BIL received. Operation [{evt.OperationId}]", evt);
```
Comment (polonzer) [inline src/Lykke.Service.Operations/Workflow/Sagas/BlockchainCashoutSaga.cs]: Isn't it excessive logging?

File: src/Lykke.Service.Operations/Workflow/Sagas/BlockchainCashoutSaga.cs [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+                    evt.TransactionHash                    
+                }.ToJson()
+            };
+
+            commandSender.SendCommand(command, "operations");
+        }
+
+        public async Task Handle(OperationExecutionFailedEvent evt, ICommandSender commandSender)
+        {
+            _log.Info($"OperationExecutionFailedEvent for BIL received. Operation [{evt.OperationId}]", evt);
```
Comment (polonzer) [inline src/Lykke.Service.Operations/Workflow/Sagas/BlockchainCashoutSaga.cs]: Isn't it excessive logging?

File: src/Lykke.Service.Operations/Workflow/Sagas/MeSaga.cs [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+            _assetsServiceWithCache = assetsServiceWithCache;
+        }
+
+        [UsedImplicitly]
+        public async Task Handle(ExternalExecutionActivityCreatedEvent evt, ICommandSender commandSender)
+        {
+            if (evt.Type == "WaitForResultsFromME")
+            {
+                // NOP. Step for waiting ME results.
+            }
```
Comment (polonzer) [inline src/Lykke.Service.Operations/Workflow/Sagas/MeSaga.cs]: Else?

File: src/Lykke.Service.Operations/Workflow/Sagas/MeSaga.cs [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+            if (evt.Type == "WaitForResultsFromME")
+            {
+                // NOP. Step for waiting ME results.
+            }            
+        }
+
+        [UsedImplicitly]
+        public async Task Handle(CashOutProcessedEvent evt, ICommandSender commandSender)
+        {
+            _log.Info($"CashOutProcessedEvent for operation [{evt.OperationId}] received", evt);
```
Comment (polonzer) [inline src/Lykke.Service.Operations/Workflow/Sagas/MeSaga.cs]: Isn't it excessive logging?

File: src/Lykke.Service.Operations/Workflow/Sagas/MeSaga.cs [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+                }.ToJson()
+            };
+
+            commandSender.SendCommand(command, "operations");
+        }
+
+        [UsedImplicitly]
+        public async Task Handle(MeCashoutFailedEvent evt, ICommandSender commandSender)
+        {
+            _log.Info($"MeCashoutFailedEvent for operation [{evt.OperationId}] received", evt);
```
Comment (polonzer) [inline src/Lykke.Service.Operations/Workflow/Sagas/MeSaga.cs]: Isn't it excessive logging?

File: src/Lykke.Service.Operations/Workflow/Sagas/WorkflowSaga.cs [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+        private ILog _log;
+
+        public WorkflowSaga(ILogFactory logFactory)
+        {
+            _log = logFactory.CreateLog(this);
+        }
+
+        public async Task Handle(OperationCreatedEvent evt, ICommandSender commandSender)
+        {
+            _log.Info($"OperationCreatedEvent for operation [{evt.Id}] received", evt);
```
Comment (polonzer) [inline src/Lykke.Service.Operations/Workflow/Sagas/WorkflowSaga.cs]: Isn't it excessive logging?

File: src/Lykke.Service.Operations/Workflow/Validation/AddressValidator.cs [hunk 19]
```diff
@@ -0,0 +1,439 @@
...
+
+                if (result == ValidationErrors.None)
+                    return true;
+            }
+
+            return false;
+        }
+    }
+
+    public interface ISrvBlockchainReader
```
Comment (polonzer) [inline src/Lykke.Service.Operations/Workflow/Validation/AddressValidator.cs]: Why so many classes in a single file?

File: tests/Lykke.Service.Operations.Tests/WorkflowTests.cs [hunk 20]
```diff
@@ -1,191 +1,256 @@
...
 
-            var result = wf.Run(operation);
+            Assert.Equal(OperationStatus.Accepted, operation.Status);            
+        }
 
-            Trace.WriteLine(JsonConvert.SerializeObject(operation, Formatting.Indented));
-            Trace.WriteLine(JsonConvert.SerializeObject(result, Formatting.Indented));
-            Assert.Equal(WorkflowState.Complete, result.State);
-            Assert.Equal(OperationStatus.Accepted, operation.Status);
+        //[Theory]
```
Comment (polonzer) [inline tests/Lykke.Service.Operations.Tests/WorkflowTests.cs]: Why is this commented?

File: tests/Lykke.Service.Operations.Tests/WorkflowTests.cs [hunk 21]
```diff
@@ -1,191 +1,256 @@
...
+                Trace.WriteLine(result.ToJson());
+                return;
+            }
 
-            operation.Create(Guid.NewGuid(), new Guid("27fe9c28-a18b-4939-8ebf-a70061fbfa05"), type, inputValues);                        
+            //Trace.WriteLine(JsonConvert.SerializeObject(operation, Formatting.Indented));
+            //Trace.WriteLine("--- WF result ---");
+            //Trace.WriteLine(JsonConvert.SerializeObject(result, Formatting.Indented));
+            
+            //await wfService.CompleteActivity(operation, operation.GetConfirmationActivity().ActivityId, JObject.FromObject(new { Confirmation = "111" }));
```
Comment (polonzer) [inline tests/Lykke.Service.Operations.Tests/WorkflowTests.cs]: Why is this commented?

File: src/Lykke.Service.Operations/Workflow/Sagas/MeSaga.cs (hunk 15 above)
Comment (mmusin) [reply src/Lykke.Service.Operations/Workflow/Sagas/MeSaga.cs]: nothing :)

File: src/Lykke.Service.Operations/Workflow/Validation/AddressValidator.cs (hunk 19 above)
Comment (mmusin) [reply src/Lykke.Service.Operations/Workflow/Validation/AddressValidator.cs]: legacy from LW, will be removed soon

File: tests/Lykke.Service.Operations.Tests/WorkflowTests.cs (hunk 20 above)
Comment (mmusin) [reply tests/Lykke.Service.Operations.Tests/WorkflowTests.cs]: don't remember :( it's integration tests, draft version

File: src/Lykke.Service.Operations/Workflow/Commands/ExecuteOperationCommand.cs (hunk 9 above)
Comment (mmusin) [reply src/Lykke.Service.Operations/Workflow/Commands/ExecuteOperationCommand.cs]: Yes, used messagepack by default in cqrs-module

**Pull request merged**